lib/modules/asset/AssetService.ts

```diff
@@ -268,6 +274,201 @@ export class AssetService {
     return result;
   }
 
+  public async migrateTenant(
+    user: User,
+    assetsList: string[],
+    engineId: string,
+    newEngineId: string
+  ) {
+    return lock(`engine:${engineId}:${newEngineId}`, async () => {
+      const recovery = new RecoveryQueue();
+
+      if (!user.profileIds.includes("admin")) {
+        throw new BadRequestError(
+          `User ${user._id} is not authorized to migrate assets`
+        );
+      }
+
+      try {
+        // check if tenant destination of the the same group
+        const engine = await this.getEngine(engineId);
+        const newEngine = await this.getEngine(newEngineId);
+
+        if (engine.group !== newEngine.group) {
+          throw new BadRequestError(
+            `Tenant ${newEngineId} is not in the same group as ${engineId}`
+          );
+        }
+
+        const assets = await this.sdk.document.mGet<AssetContent>(
+          engineId,
+          InternalCollection.ASSETS,
+          assetsList
+        );
+
+        // check if the assets exists in the other engine
+        const existingAssets = await this.sdk.document.mGet<AssetContent>(
+          newEngineId,
+          InternalCollection.ASSETS,
+          assetsList
+        );
+
+        if (existingAssets.successes.length > 0) {
+          throw new BadRequestError(
+            `Assets ${existingAssets.successes
+              .map((asset) => asset._id)
+              .join(", ")} already exists in engine ${newEngineId}`
+          );
+        }
+        const assetsToMigrate = assets.successes.map((asset) => ({
+          _id: asset._id,
+          body: asset._source,
+        }));
+
+        const devices = await this.sdk.document.search<AssetContent>(
+          engineId,
+          InternalCollection.DEVICES,
+          {
+            query: {
+              bool: {
+                filter: {
+                  terms: {
+                    assetId: assetsList,
+                  },
+                },
+              },
+            },
+          }
+        );
+
+        // Map linked devices for assets.
+        const assetLinkedDevices = assets.successes
+          .filter((asset) => asset._source.linkedDevices.length > 0)
+          .map((asset) => ({
+            assetId: asset._id,
+            linkedDevices: asset._source.linkedDevices,
+          }));
+
+        // Extra recovery step to relink back assets to their devices in case of rollback
+        recovery.addRecovery(async () => {
+          // Link the devices to the new assets
+          for (const asset of assetLinkedDevices) {
+            const assetId = asset.assetId;
+            for (const device of asset.linkedDevices) {
+              const deviceId = device._id;
+              const measureNames = device.measureNames;
+              await ask<AskDeviceLinkAsset>(
```

Comment:
You should populate an array with promise and resolve them with promise.all, running await here will do them one by one. running promise will create a batch running them in // it will be much faster